code/modules/mob/living/silicon/robot/robot.dm

```diff
@@ -766,15 +767,20 @@
 	if(loc == src)
 		return
 	old_upgrade.deactivate(src)
-	upgrades -= old_upgrade
+	for(var/obj/item/borg/upgrade/current_upgrade in src.upgrades)
+		if(istype(current_upgrade, old_upgrade))
+			src.upgrades.Remove(current_upgrade)
```

Comment:
You iterate all these but never call deactivate or emit signals for these. Explain.